File: app/viewmodel/UpdateChecker.cpp
#include "viewmodel/UpdateChecker.h"

#include <QCoreApplication>
#include <QJsonArray>
#include <QJsonDocument>
#include <QJsonObject>
#include <QJsonParseError>
#include <QNetworkAccessManager>
#include <QNetworkReply>
#include <QNetworkRequest>
#include <QRegularExpression>
#include <QtGlobal>

namespace fluent::gallery {
namespace {
constexpr const char* kLatestReleaseApi =
    "https://api.github.com/repos/calvinhxx/Fluent-Qt/releases/latest";

bool assetNameMatchesPlatform(const QString& assetName,
                              const QString& platformKey) {
    const QString name = assetName.toLower();

    if (name.contains(QStringLiteral(".sha256"))) {
        return false;
    }

    if (platformKey == QStringLiteral("windows-arm64")) {
        return name.contains(QStringLiteral("windows-arm64")) &&
               name.endsWith(QStringLiteral(".exe"));
    }

    if (platformKey == QStringLiteral("windows-x64")) {
        return name.contains(QStringLiteral("windows-x64")) &&
               name.endsWith(QStringLiteral(".exe")) &&
               !name.contains(QStringLiteral("qt5.15"));
    }

    if (platformKey == QStringLiteral("macos-arm64")) {
        return name.contains(QStringLiteral("darwin-arm64")) &&
               name.endsWith(QStringLiteral(".dmg"));
    }

    if (platformKey == QStringLiteral("macos-x64")) {
        return name.contains(QStringLiteral("darwin-x86_64")) &&
               name.endsWith(QStringLiteral(".dmg")) &&
               !name.contains(QStringLiteral("qt5.15"));
    }

    if (platformKey == QStringLiteral("linux-arm64")) {
        return name.contains(QStringLiteral("linux-arm64")) &&
               name.endsWith(QStringLiteral(".deb"));
    }

    if (platformKey == QStringLiteral("linux-x64")) {
        return name.contains(QStringLiteral("linux-x86_64")) &&
               name.endsWith(QStringLiteral(".deb"));
    }

    return false;
}
}  // namespace

UpdateChecker::UpdateChecker(QObject* parent) : QObject(parent) {
    qRegisterMetaType<fluent::gallery::UpdateChecker::Result>(
        "fluent::gallery::UpdateChecker::Result");
    m_network = new QNetworkAccessManager(this);
}

UpdateChecker::~UpdateChecker() {
    if (m_reply) {
        m_reply->abort();
        m_reply->deleteLater();
        m_reply = nullptr;
    }
}

bool UpdateChecker::isChecking() const { return m_reply != nullptr; }

QString UpdateChecker::currentVersion() const {
    const QString version = QCoreApplication::applicationVersion().trimmed();
    return version.isEmpty() ? QStringLiteral("0.0.0") : version;
}

QString UpdateChecker::platformLabel() const { return platformDisplayName(); }

void UpdateChecker::checkForUpdates() {
    if (m_reply) {
        return;
    }

    QNetworkRequest request{QUrl(QString::fromLatin1(kLatestReleaseApi))};
    request.setHeader(QNetworkRequest::UserAgentHeader,
                      QStringLiteral("Fluent-Qt-Gallery/%1")
                          .arg(currentVersion()));
    request.setRawHeader("Accept", "application/vnd.github+json");
    request.setAttribute(QNetworkRequest::RedirectPolicyAttribute,
                         QNetworkRequest::NoLessSafeRedirectPolicy);

    m_reply = m_network->get(request);
    connect(m_reply, &QNetworkReply::finished, this,
            &UpdateChecker::handleReplyFinished);
    emit checkStarted();
}

UpdateChecker::Result UpdateChecker::parseReleasePayload(
    const QByteArray& payload) const {
    Result result;
    result.currentVersion = currentVersion();

    QJsonParseError parseError;
    const QJsonDocument document = QJsonDocument::fromJson(payload, &parseError);
    if (parseError.error != QJsonParseError::NoError || !document.isObject()) {
        result.message = tr("GitHub release response could not be parsed.");
        return result;
    }

    const QJsonObject release = document.object();
    const QString tagName = release.value(QStringLiteral("tag_name")).toString();
    const QString htmlUrl = release.value(QStringLiteral("html_url")).toString();
    result.releaseName = release.value(QStringLiteral("name")).toString();
    result.latestVersion = normalizedVersion(tagName);
    result.releaseUrl = QUrl(htmlUrl);

    if (result.latestVersion.isEmpty() || !result.releaseUrl.isValid()) {
        result.message = tr("Latest release metadata is incomplete.");
        return result;
    }

    const int comparison =
        compareVersions(result.latestVersion, result.currentVersion);
    if (comparison <= 0) {
        result.status = Result::Status::UpToDate;
        result.message = tr("You are on the latest version (%1).")
                             .arg(result.currentVersion);
        return result;
    }

    const QJsonArray assets = release.value(QStringLiteral("assets")).toArray();
    result.assetUrl = selectPlatformAsset(assets, &result.assetName);
    result.status = Result::Status::UpdateAvailable;

    if (result.assetUrl.isValid()) {
        result.message =
            tr("Version %1 is available for %2.")
                .arg(result.latestVersion, platformDisplayName());
    } else {
        result.message =
            tr("Version %1 is available. Open the release page to download it.")
                .arg(result.latestVersion);
    }

    return result;
}

QUrl UpdateChecker::selectPlatformAsset(const QJsonArray& assets,
                                        QString* assetName) const {
    const QString key = platformKey();
    for (const QJsonValue& value : assets) {
        if (!value.isObject()) {
            continue;
        }

        const QJsonObject asset = value.toObject();
        const QString name = asset.value(QStringLiteral("name")).toString();
        if (!assetNameMatchesPlatform(name, key)) {
            continue;
        }

        const QUrl url{
            asset.value(QStringLiteral("browser_download_url")).toString()};
        if (!url.isValid()) {
            continue;
        }

        if (assetName) {
            *assetName = name;
        }
        return url;
    }

    if (assetName) {
        assetName->clear();
    }
    return {};
}

void UpdateChecker::handleReplyFinished() {
    QNetworkReply* reply = m_reply;
    m_reply = nullptr;

    Result result;
    result.currentVersion = currentVersion();

    const QVariant statusCode =
        reply->attribute(QNetworkRequest::HttpStatusCodeAttribute);
    const int httpStatus = statusCode.isValid() ? statusCode.toInt() : 0;
    if (reply->error() != QNetworkReply::NoError) {
        result.message = reply->errorString();
    } else if (httpStatus >= 400) {
        result.message =
            tr("GitHub release request failed with HTTP %1.").arg(httpStatus);
    } else {
        result = parseReleasePayload(reply->readAll());
    }

    reply->deleteLater();
    emit checkFinished(result);
}

int UpdateChecker::compareVersions(const QString& left, const QString& right) {
    const QStringList leftParts = normalizedVersion(left).split('.');
    const QStringList rightParts = normalizedVersion(right).split('.');

    for (int index = 0; index < 3; ++index) {
        const int leftValue =
            index < leftParts.size() ? leftParts.at(index).toInt() : 0;
        const int rightValue =
            index < rightParts.size() ? rightParts.at(index).toInt() : 0;

        if (leftValue < rightValue) {
            return -1;
        }
        if (leftValue > rightValue) {
            return 1;
        }
    }

    return 0;
}

QString UpdateChecker::normalizedVersion(const QString& version) {
    QString normalized = version.trimmed();
    if (normalized.startsWith(QLatin1Char('v')) ||
        normalized.startsWith(QLatin1Char('V'))) {
        normalized.remove(0, 1);
    }

    const int suffixIndex = normalized.indexOf(QLatin1Char('-'));
    if (suffixIndex >= 0) {
        normalized.truncate(suffixIndex);
    }

    static const QRegularExpression validVersionPattern{
        QStringLiteral(R"(^\d+(?:\.\d+){0,2})")};
    const QRegularExpressionMatch match =
        validVersionPattern.match(normalized);
    return match.hasMatch() ? match.captured(0) : QString();
}

QString UpdateChecker::platformKey() {
#if defined(Q_OS_WIN)
#if defined(Q_PROCESSOR_ARM_64)
    return QStringLiteral("windows-arm64");
#else
    return QStringLiteral("windows-x64");
#endif
#elif defined(Q_OS_MACOS) || defined(Q_OS_MAC)
#if defined(Q_PROCESSOR_ARM_64)
    return QStringLiteral("macos-arm64");
#else
    return QStringLiteral("macos-x64");
#endif
#elif defined(Q_OS_LINUX)
#if defined(Q_PROCESSOR_ARM_64)
    return QStringLiteral("linux-arm64");
#else
    return QStringLiteral("linux-x64");
#endif
#else
    return QStringLiteral("release-page");
#endif
}

QString UpdateChecker::platformDisplayName() {
    const QString key = platformKey();
    if (key == QStringLiteral("windows-arm64")) {
        return QStringLiteral("Windows ARM64");
    }
    if (key == QStringLiteral("windows-x64")) {
        return QStringLiteral("Windows x64");
    }
    if (key == QStringLiteral("macos-arm64")) {
        return QStringLiteral("macOS Apple Silicon");
    }
    if (key == QStringLiteral("macos-x64")) {
        return QStringLiteral("macOS Intel");
    }
    if (key == QStringLiteral("linux-arm64")) {
        return QStringLiteral("Linux ARM64");
    }
    if (key == QStringLiteral("linux-x64")) {
        return QStringLiteral("Linux x64");
    }
    return QStringLiteral("this platform");
}

}  // namespace fluent::gallery

File: bindings/pyside6/src/fluentqt/gallery/update_checker.py
"""GitHub release checker used by the Python Gallery settings page."""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum
import json
import platform
import re

import fluentqt
from PySide6.QtCore import QObject, QUrl, Signal
from PySide6.QtNetwork import (
    QNetworkAccessManager,
    QNetworkReply,
    QNetworkRequest,
)


_LATEST_RELEASE_API = (
    "https://api.github.com/repos/calvinhxx/Fluent-Qt/releases/latest"
)


class UpdateStatus(Enum):
    UpToDate = "up-to-date"
    UpdateAvailable = "update-available"
    Error = "error"


@dataclass
class UpdateResult:
    status: UpdateStatus = UpdateStatus.Error
    current_version: str = ""
    latest_version: str = ""
    release_name: str = ""
    release_url: QUrl | None = None
    asset_url: QUrl | None = None
    asset_name: str = ""
    message: str = ""


def normalized_version(version: str) -> str:
    normalized = version.strip()
    if normalized[:1].lower() == "v":
        normalized = normalized[1:]
    normalized = normalized.split("-", 1)[0]
    match = re.match(r"^\d+(?:\.\d+){0,2}", normalized)
    return match.group(0) if match else ""


def compare_versions(left: str, right: str) -> int:
    def parts(value: str) -> list[int]:
        normalized = normalized_version(value)
        return [int(part) for part in normalized.split(".") if part]

    left_parts = parts(left)
    right_parts = parts(right)
    for index in range(3):
        left_value = left_parts[index] if index < len(left_parts) else 0
        right_value = right_parts[index] if index < len(right_parts) else 0
        if left_value != right_value:
            return -1 if left_value < right_value else 1
    return 0


def platform_key() -> str:
    machine = platform.machine().lower()
    arm64 = "arm" in machine or "aarch64" in machine
    system = platform.system()
    if system == "Windows":
        return "windows-arm64" if arm64 else "windows-x64"
    if system == "Darwin":
        return "macos-arm64" if arm64 else "macos-x64"
    if system == "Linux":
        return "linux-arm64" if arm64 else "linux-x64"
    return "release-page"


def platform_display_name() -> str:
    return {
        "windows-arm64": "Windows ARM64",
        "windows-x64": "Windows x64",
        "macos-arm64": "macOS Apple Silicon",
        "macos-x64": "macOS Intel",
        "linux-arm64": "Linux ARM64",
        "linux-x64": "Linux x64",
    }.get(platform_key(), "this platform")


def _asset_matches_platform(name: str, key: str) -> bool:
    folded = name.lower()
    if ".sha256" in folded:
        return False
    if key == "windows-arm64":
        return "windows-arm64" in folded and folded.endswith(".exe")
    if key == "windows-x64":
        return (
            "windows-x64" in folded
            and folded.endswith(".exe")
            and "qt5.15" not in folded
        )
    if key == "macos-arm64":
        return "darwin-arm64" in folded and folded.endswith(".dmg")
    if key == "macos-x64":
        return (
            "darwin-x86_64" in folded
            and folded.endswith(".dmg")
            and "qt5.15" not in folded
        )
    if key == "linux-arm64":
        return "linux-arm64" in folded and folded.endswith(".deb")
    if key == "linux-x64":
        return "linux-x86_64" in folded and folded.endswith(".deb")
    return False


class GalleryUpdateChecker(QObject):
    """Python equivalent of the C++ Gallery UpdateChecker."""

    checkStarted = Signal()
    checkFinished = Signal(object)

    def __init__(self, parent: QObject | None = None) -> None:
        super().__init__(parent)
        self._network = QNetworkAccessManager(self)
        self._reply: QNetworkReply | None = None

    def is_checking(self) -> bool:
        return self._reply is not None

    def current_version(self) -> str:
        version = str(fluentqt.__version__).strip()
        return version or "0.0.0"

    def platform_label(self) -> str:
        return platform_display_name()

    def check_for_updates(self) -> None:
        if self._reply is not None:
            return
        request = QNetworkRequest(QUrl(_LATEST_RELEASE_API))
        request.setHeader(
            QNetworkRequest.UserAgentHeader,
            "Fluent-Qt-Gallery/{0}".format(self.current_version()),
        )
        request.setRawHeader(b"Accept", b"application/vnd.github+json")
        request.setAttribute(
            QNetworkRequest.RedirectPolicyAttribute,
            QNetworkRequest.NoLessSafeRedirectPolicy,
        )
        self._reply = self._network.get(request)
        self._reply.finished.connect(self._handle_reply_finished)
        self.checkStarted.emit()

    def parse_release_payload(self, payload: bytes) -> UpdateResult:
        result = UpdateResult(current_version=self.current_version())
        try:
            release = json.loads(payload.decode("utf-8"))
        except (UnicodeDecodeError, json.JSONDecodeError):
            result.message = "GitHub release response could not be parsed."
            return result
        if not isinstance(release, dict):
            result.message = "GitHub release response could not be parsed."
            return result

        result.latest_version = normalized_version(str(release.get("tag_name", "")))
        result.release_name = str(release.get("name", ""))
        result.release_url = QUrl(str(release.get("html_url", "")))
        if not result.latest_version or not result.release_url.isValid():
            result.message = "Latest release metadata is incomplete."
            return result

        if compare_versions(result.latest_version, result.current_version) <= 0:
            result.status = UpdateStatus.UpToDate
            result.message = "You are on the latest version ({0}).".format(
                result.current_version
            )
            return result

        key = platform_key()
        for asset in release.get("assets", []):
            if not isinstance(asset, dict):
                continue
            name = str(asset.get("name", ""))
            if not _asset_matches_platform(name, key):
                continue
            url = QUrl(str(asset.get("browser_download_url", "")))
            if url.isValid():
                result.asset_name = name
                result.asset_url = url
                break

        result.status = UpdateStatus.UpdateAvailable
        if result.asset_url is not None and result.asset_url.isValid():
            result.message = "Version {0} is available for {1}.".format(
                result.latest_version, self.platform_label()
            )
        else:
            result.message = (
                "Version {0} is available. Open the release page to download it."
            ).format(result.latest_version)
        return result

    def _handle_reply_finished(self) -> None:
        reply = self._reply
        self._reply = None
        if reply is None:
            return

        result = UpdateResult(current_version=self.current_version())
        status = reply.attribute(QNetworkRequest.HttpStatusCodeAttribute)
        http_status = int(status) if status is not None else 0
        if reply.error() != QNetworkReply.NoError:
            result.message = reply.errorString()
        elif http_status >= 400:
            result.message = (
                "GitHub release request failed with HTTP {0}.".format(http_status)
            )
        else:
            result = self.parse_release_payload(bytes(reply.readAll()))
        reply.deleteLater()
        self.checkFinished.emit(result)


__all__ = [
    "GalleryUpdateChecker",
    "UpdateResult",
    "UpdateStatus",
    "compare_versions",
    "normalized_version",
    "platform_display_name",
    "platform_key",
]
